test/sqllogictest/joins.slt

```diff
@@ -602,3 +602,92 @@ CREATE TABLE join_fail (la date);
 
 query error NATURAL/USING join column "la" cannot be cast to uniform type: integer vs date
 SELECT la FROM l JOIN join_fail USING (la)
+
```

Comment:
This query returns a wrong result in this branch, and works correctly with the previous commit:

```
EXPLAIN SELECT * FROM t1 LEFT JOIN t2 ON t2.f1 IS NULL WHERE t1.f1 + 3 = 6 AND t1.f1 + t2.f2 IS NULL;

                  Optimized Plan                  
--------------------------------------------------
 %0 =                                            +
 | Get materialize.public.t1 (u57)               +
 | Filter (6 = (#0 + 3))                         +
 | ArrangeBy ()                                  +
                                                 +
 %1 =                                            +
 | Get materialize.public.t2 (u61)               +
 | Filter isnull(#0)                             +
                                                 +
 %2 = Let l0 =                                   +
 | Join %0 %1                                    +
 | | implementation = Differential %1 %0.()      +
 | | demand = (#0..#3)                           +
                                                 +
 %3 =                                            +
 | Get %2 (l0)                                   +
 | Filter isnull(#0), (6 = (#0 + 3))             +
                                                 +
 %4 =                                            +
 | Get %2 (l0)                                   +
 | Filter (6 = (#0 + 3))                         +
 | Distinct group=(#0, #1)                       +
 | Negate                                        +
                                                 +
 %5 =                                            +
 | Get materialize.public.t1 (u57)               +
 | Filter (6 = (#0 + 3))                         +
 | Distinct group=(#0, #1)                       +
                                                 +
 %6 =                                            +
 | Union %4 %5                                   +
                                                 +
 %7 =                                            +
 | Get materialize.public.t1 (u57)               +
 | ArrangeBy (#1, #0)                            +
                                                 +
 %8 =                                            +
 | Join %6 %7 (= #0 #2) (= #1 #3)                +
 | | implementation = Differential %6 %7.(#1, #0)+
 | | demand = (#0, #1)                           +
 | Filter (6 = (#0 + 3))                         +
 | Map null, null                                +
 | Project (#0, #1, #4, #5)                      +
                                                 +
 %9 =                                            +
 | Union %3 %8                                   +
```

The difference with respect to the previous commit is ```Filter (6 = (#0 + 3))``` in #0 . For some reason I can not possibly fanthom, this produces a fanthom row:

```
materialize=> SELECT * FROM t1 LEFT JOIN t2 ON t2.f1 IS NULL WHERE t1.f1 + 3 = 6 AND t1.f1 + t2.f2 IS NULL;
 f1 | f2 | f1 | f2 
----+----+----+----
  3 |  3 |    |                    <- Wrong result
(1 row)
```

The same query without a ```WHERE``` clause shows that this row does not exist in the join product:
````
materialize=> SELECT * FROM t1 LEFT JOIN t2 ON t2.f1 IS NULL ;
 f1 | f2 | f1 | f2 
----+----+----+----
    |  0 |    |  0
  2 |  2 |    |  0
  3 |  3 |    |  0 <- There is no row (3,3,NULL, NULL)
  4 |  4 |    |  0
  5 |  5 |    |  0
  6 |  6 |    |  0
  7 |  7 |    |  0
  8 |  8 |    |  0
  9 |  9 |    |  0
  1 |  1 |    |  0
```

Schema:
```
CREATE TABLE t1 (f1 DOUBLE PRECISION, f2 DOUBLE PRECISION NOT NULL);
CREATE INDEX t1i1 ON t1(f1);
CREATE INDEX t1i2 ON t1(f2, f1);

INSERT INTO t1 VALUES (NULL, 0);

INSERT INTO t1 VALUES (1, 1);
INSERT INTO t1 VALUES (2, 2);
INSERT INTO t1 VALUES (3, 3);
INSERT INTO t1 VALUES (4, 4);
INSERT INTO t1 VALUES (5, 5);
INSERT INTO t1 VALUES (6, 6);
INSERT INTO t1 VALUES (7, 7);
INSERT INTO t1 VALUES (8, 8);
INSERT INTO t1 VALUES (9, 9);

CREATE TABLE t2 (f1 DOUBLE PRECISION, f2 DOUBLE PRECISION NOT NULL);
CREATE INDEX t2i1 ON t2(f1);
CREATE INDEX i2i2 ON t2(f2, f1);

INSERT INTO t2 VALUES (NULL, 0);

INSERT INTO t2 VALUES (1, 1);
INSERT INTO t2 VALUES (2, 2);
INSERT INTO t2 VALUES (3, 3);
INSERT INTO t2 VALUES (4, 4);
INSERT INTO t2 VALUES (5, 5);
INSERT INTO t2 VALUES (6, 6);
INSERT INTO t2 VALUES (7, 7);
INSERT INTO t2 VALUES (8, 8);
INSERT INTO t2 VALUES (9, 9);
```

Reply:
Here is another query that exhibits a similar symthom:

```
SELECT  * FROM t2  AS a1  LEFT  JOIN t1  AS a2  ON ( a2 .f1  > a1 .f2  )  WHERE a2 .f1  IS  NULL  AND a1  IS  NOT  NULL  ;```

Though in this case the extra filter looks like this ```Filter !(isnull(record_create(#0, #1))) ````